osu.Game/Rulesets/UI/ReplayRecorder.cs

```diff
@@ -55,7 +55,9 @@ protected override void LoadComplete()
         protected override void Dispose(bool isDisposing)
         {
             base.Dispose(isDisposing);
-            spectatorClient?.EndPlaying();
+
+            if (spectatorClient != null && gameplayState != null)
```

Comment:
Is the problem the nullability or what?

Reply:
Yeah, just on an initial read it seems like the `EndPlaying` call may be incorrectly forgotten when there's no `gameplayState`, I think it would be worthwhile explaining what the null case means here.

Reply:
I've simplified the code without adding a comment. I don't know how to explain this because it's the same `LoadComplete`/`Disposal` pattern we use everywhere else so it seems redundant. If we do this then I assert we must comment the same in every place including everything that uses Realm subscriptions, for example.

Reply:
My specific issue was with the *parameter* here, not the `spectatorClient` null check (which I agree we use all over the place and have no issue with). Reading further I can see it's due to it being a resolved dependency, but it feels a bit weird structurally to be resolving `GameplayState` for the single purpose of passing to `SpectatorClient`. That's an issue in `master` too though.

Reply:
Are you okay with the current implementation then?

Reply:
Yeah I think it's okay for now.